packages/clay-date-picker/src/index.tsx

```diff
@@ -27,6 +32,39 @@ import {FirstDayOfWeek, IAriaLabels, IYears} from './types';
 
 export {FirstDayOfWeek};
 
+function isYearWithinYears(year: number, years: IYears) {
+	return years.start <= year && year <= years.end;
+}
+
+export const RANGE_SEPARATOR = ' - ';
+
+function fromStringToRange(
+	value: string,
+	dateFormat: string,
+	referenceDate: Date
+): readonly [Date, Date] {
+	const [fromDateString, toDateString] = value.split(RANGE_SEPARATOR);
+
+	return [
+		parseDate(fromDateString, dateFormat, referenceDate),
+		toDateString
+			? parseDate(toDateString, dateFormat, referenceDate)
+			: referenceDate,
+	];
+}
+
+function fromRangeToString(
+	range: readonly [Date, Date],
+	dateFormat: string
+) {
+	const [startDate, endDate] = range;
+
+	return `${formatDate(startDate, dateFormat)}${RANGE_SEPARATOR}${formatDate(
+		endDate,
+		dateFormat
+	)}`;
+}
```

Comment:
Although I have a preference to keep it in the helpers, I have the feeling that the component is more bloated with information, the separation of files helps me to read better as opposed to harming at least for me. I have a preference to keep this below the interface and to consider the interface declaration first in the component file, it is easier to know what the component expects.

cc @wincent

Reply:
There are arguments for and against, but I come down on the side of preferring locality (or colocation, or whatever you want to call it). Here are some reasons:

- If you put it in a separate file, the user has to jump to another place to look at it.
- If all you want to do is get it "out of the way", you can move it to the bottom of the same file (because functions are all "hoisted" in JS; ie. usable in code that appears above their definition); this way, a person landing in the file can still see the important stuff (ie. the component) at the top.
- Having it in a separate file means having to `import` it, which clutters the header area of the file with more boilerplate.
- Keeping it in the same file means that you can rename, move this file and know this internal function that it needs will always go along for the ride. Conversely, if you extract functions out into another file, every time you want to move that other file, you have to update all the imports. In other words, splitting stuff into more files adds friction and makes iteration/evolution slower.
- Putting it in a separate file means that anybody looking at the function cannot change it without first searching for an analysis all usage sites (eg. they can't rename it, they can't change the behavior etc); conversely, if the function is in the same file as the only user, it is easy to rename/refactor, and contemplate other behavior changes. Another way of saying this is that the cognitive and adminisrative load of reading and understanding an _internal_ function is always going to be lower than that of an _exported_ function.
- File size is a weak argument for wanting to extract functions; modern editors and computers can easily handle large files, and provide good tools for visualizing and organizing them (eg. code-folding, split panes etc).

And even though all of those arguments apply to functions, they resonate with a broader trend in the industry to recognize the value of colocation in general: GraphQL libraries colocate query fragments with the components that need the corresponding data; CSS-in-JS libraries colocate style definitions inside the components that need those styles; even JSX itself is an example (remember the "rethink best practices" meme? That was all about saying that what we had been calling "separation of concerns" was actually "separation of technologies", so splitting stuff into JS + HTML + CSS was wrong, and a better unit of abstraction and encapsulation was the "component", which is why we should "mix" together presentational, structural, and behavioral concerns in the same module, redefining what we consider to be a "concern"... that's actual an argument for colocation in disguise).

Obviously, it's up to you how you divide up your code and none of this is blocking feedback. I just honestly consider this to be good advice and want to share it. Similar arguments have been made by people much smarter than me (eg. [Carmack, here](https://twitter.com/id_aa_carmack/status/1333605572617363458)):

> A long time ago, I wrote about how a big function can be better than many small ones: http://number-none.com/blow/john_carmack_on_inlined_code.html I think it also applies to processes in a system.  I just said this: "A million line codebase is almost cozy today; I think we could combine many things and benefit."

So, at the surface level, he's talking about distrubuted systems. But what he's really talking about is colocation.

Reply:
@wincent Yeah, these are great arguments, it's a different perspective from that point of view looking at how components "should" be structured so as not to require separation of technology or just to bring things together in the same file, I agree that this reduced the cognitive load for make changes.

Well, about large files it is really quite debatable or swell with a lot of information to put the intrinsic complexity under the carpet also a point but it is also quite relative and varies from case to case, for example, during migration I saw file with minimum 2 thousand lines, some 3 thousand and a large application in just one component that made "millions" of rules, so I would consider this a large file and had an explicit high complexity, all the things related to the component/app were there, that of course imposes other problems that are not related to the locality but as the component was created without responsibility what I saw myself wasting hours trying to understand to discover the edge cases, in the end give up trying to understand why it was Metal.js and went with a solution of zero, this issue doesn’t really matter, but the result was breaking the component into small pieces with responsibilities where each small component could be composed and reused in other places, so in the end it became about 10 files which makes it easier to understand part of the application and makes it easy to reason what’s happening and helps to discover flaws more quickly... although locality is not imposed here I mean that this is more a matter of balance and depends a lot on the case and how it is done.

For some cases allocating everything can mean adding complexity or just keeping everything close by with something very structured, like, let's keep HTML + CSS + JavaScript close, it took us a while to add it all close because it required us to rethink the shape of everyone together, PHP started a part of this but quickly started to generate problems at scale, then we started to evolve to the frameworks, without considering MVC here..., on the JavaScript side, some find Vue interesting for doing this, uniting everything in the same file, some find that can get complex, so we have JSX + CSS in JS + JS in the same file, they are different technologies but they all converge to JS that approximate the same corresponding technology, this means that to understand and benefit from the locality we would need to structure how each technology it would be structured within the same file or how to achieve this without getting complex to understand or scale. I hope you understand the point of view that I am raising here.

Bringing it to our reality in Clay, we can sometimes think that 500 lines or 200 is something very big, depending on the point of view and the component maybe it is something very big or maybe not, I think that should not be taken into account so much for to say if we should take all the functions and put it on the same file where it is being used but we should in a way evaluate if it makes sense in the proposal: DatePicker `index.js` needs to know this, it just concentrates the composition of all components, assembling the blocks... well I understood the objectives of the locality but I see that there must always be a scale and it varies a lot from case to case I don’t have the cake recipe of when we should follow this against the locality so I won’t be able to argue much here.

Thank you for bringing this point of view!

Reply:
Yeah, the answer is always "it depends". One has to be careful about transferring one's experiences from 2,000 or 3,000 line "nightmare" files in a project that (from what I've heard) had really tied itself into knots, into different contexts.

You can easily find counter examples where well-written multi-thousand-line files have worked out brilliantly... (look in the Git project, or the Linux kernel, or TypeScript etc). Size alone should _not_ be the criteria for judging whether something is "maintainable" or not.

Reply:
Yes, I totally agree. I am not bringing this up as being big will be bad but it really depends, just so that other people when they see it try to take into account the points before going with locality or not.

As I had commented to get to the point of having everything together without achieving what was done in PHP at the beginning, it demanded that we do something well structured, so for sure this comes to the point here of taking as examples to reach a healthy locality or not. Considering the same ones you mentioned, Linux Kernel, TypeScript, At first, even the codebase in Go try to do this as good practice, I'm not sure... (Or just by limitation)